packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -0,0 +1,338 @@
+                    ),
+                    Models (ECInstanceId) AS (
+                      SELECT s.ECInstanceId as \`ECInstanceId\` FROM BisCore.Model s
+                      WHERE InVirtualSet(?, ECInstanceId)
+                    )
+                    SELECT r.TargetECInstanceId as \`ECInstanceId\` FROM ChildSubjects s
+                      JOIN BisCore.ElementOwnsChildElements r ON r.SourceECInstanceId = s.ECInstanceId
+                      WHERE r.TargetECClassId IS (BisCore.PhysicalPartition)
+                    UNION
+                    SELECT DISTINCT ECInstanceId FROM Models;`;
+
+    return this.executeQuery(query, bindings);
+  }
+
+  private async getHilitedSubCategories(map: Map<InstanceIdType, string[]>): Promise<string[]> {
+    const subCategoryKeys = map.get("subCategory") ?? [];
+    const categoryKeys = map.get("category") ?? [];
+
+    if (subCategoryKeys.length === 0 && categoryKeys.length === 0) {
+      return [];
+    }
+
+    const bindings: ECSqlBinding[] = [
+      { type: "idset", value: categoryKeys },
+      { type: "idset", value: subCategoryKeys },
+    ];
+    const query = `WITH
+                    CategorySubCategories (ECInstanceId) AS (
+                      SELECT r.TargetECInstanceId as \`ECInstanceId\` FROM BisCore.Category s
+                        JOIN BisCore.CategoryOwnsSubCategories r ON r.SourceECInstanceId = s.ECInstanceId
+                      WHERE r.TargetECClassId IS (BisCore.SubCategory) AND InVirtualSet(?, s.ECInstanceId)
+                    ),
+                    SubCategories (ECInstanceId) AS (
+                      SELECT s.ECInstanceId as \`ECInstanceId\` FROM BisCore.SubCategory s
+                      WHERE InVirtualSet(?, s.ECInstanceId)
+                    )
+                   SELECT DISTINCT ECInstanceId FROM CategorySubCategories
+                   UNION
+                   SELECT DISTINCT ECInstanceId FROM SubCategories;`;
+
+    return this.executeQuery(query, bindings);
+  }
+
+  private async getHilitedElements(map: Map<InstanceIdType, string[]>): Promise<string[]> {
+    const groupInformationElementKeys = map.get("groupInformationElement") ?? [];
+    let geometricElementKeySet = new Set(map.get("geometricElement") ?? []);
+    const elementKeys = map.get("element") ?? [];
+
+    if (this._hasFunctionalSchema) {
+      const functionalGeometricElements = new Set(await this.getHilitedFunctionalElements(map));
+      geometricElementKeySet = new Set([...geometricElementKeySet, ...functionalGeometricElements]);
+    }
+
+    const geometricElementKeys = Array.from(geometricElementKeySet);
+
+    if (groupInformationElementKeys.length === 0 && geometricElementKeys.length === 0 && elementKeys.length === 0) {
+      return [];
+    }
+
+    const bindings: ECSqlBinding[] = [
+      { type: "idset", value: groupInformationElementKeys },
+      { type: "idset", value: elementKeys },
+      { type: "idset", value: geometricElementKeys },
+    ];
```

Comment:
`InVirtualSet` is inefficient to use on a table that is likely to have many rows. Elements' table is one such table - we should prefer `IN` operator unless there are too many IDs (e.g. less than 1000 IDs - use `IN`, otherwise fallback to `InVirtualSet`).